### PR TITLE
trace manifest validation script 

### PR DESCRIPTION
**Pull request by ctr-martemovTT**
### PR Category

This PR implements

- https://github.com/tenstorrent/tt-metal/issues/36775
- https://github.com/tenstorrent/tt-metal/issues/36774

Implementation of the trace manifest validator for https://github.com/tenstorrent/tt-metal/issues/35838

Additionally, this PR includes pre-existing scripts attached to https://github.com/tenstorrent/tt-metal/issues/35838 -- the scripts were added in order to make the current E2E example in the readme reproducible, the focus of this PR to implement the validation for phase 2. 

```bash
# Phase 1: record ops + tensors (manifest written to bringup/artifacts/phase1/manifest.json)
python tools/bringup/phase1_record_ops.py --input-shape 1 3 256 256

# Validate the manifest before running tests
python tools/bringup/validate_trace_manifest.py --manifest bringup/artifacts/phase1/manifest.json

# Phase 2: generate a pytest file from the manifest. The generated test imports
# tracer_test_harness from tools/bringup via sys.path (no files are copied).
python tools/bringup/phase2_generate_tests.py \
    --manifest bringup/artifacts/phase1/manifest.json \
    --out-test bringup/artifacts/phase2/test_automatic_execution.py

# Phase 2: execute tests
python -m pytest --capture=fd --log-level=DEBUG --tb=short bringup/artifacts/phase2/test_automatic_execution.py
```

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:martemov/trace-manifest-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:martemov/trace-manifest-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:martemov/trace-manifest-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:martemov/trace-manifest-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:martemov/trace-manifest-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:martemov/trace-manifest-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:martemov/trace-manifest-validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:martemov/trace-manifest-validation)

### CODE REVIEW

**Review by blozano-tt — APPROVED** [review]

File: tools/bringup/tracer_op_specs.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+    )
+
+
+def load_manifest(manifest_path: Any) -> List[Record]:
+    """Load a manifest JSON file into a list of ``Record`` objects."""
+    # Confine the (possibly user-supplied) manifest path to the repo root before
+    # reading it, to prevent path traversal outside the intended scope.
+    base = str(_REPO_ROOT)
+    resolved = os.path.abspath(os.path.join(base, os.fspath(manifest_path)))
+    if not resolved.startswith(base):
```
Comment (blozano-tt) [inline tools/bringup/tracer_op_specs.py]: @ctr-martemovTT The path confinement here uses `resolved.startswith(base)`, which is a prefix bug. A sibling directory named `tt-metal-secrets` will pass the check and escape the repo root. The same PR already has a correct helper (`resolve_within_repo` using `os.path.commonpath`). Please route all path checks through that helper, or at minimum append `os.sep` to the base string.

File: tools/bringup/phase2_generate_tests.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+def _resolve_within_repo(user_path: Any, *, label: str) -> str:
+    """Resolve ``user_path`` against the repo root and confine it there.
+
+    Uses an absolute-path + ``startswith`` check against ``_REPO_ROOT`` so that
+    dynamic (CLI/user) input cannot escape the repo tree via traversal or an
+    out-of-tree absolute path.
+    """
+    resolved = os.path.abspath(os.path.join(_REPO_ROOT, os.fspath(user_path)))
+    if not resolved.startswith(_REPO_ROOT):
```
Comment (blozano-tt) [inline tools/bringup/phase2_generate_tests.py]: Same `startswith` prefix bug as above. Use `resolve_within_repo` from `tracer_op_specs` so path confinement is consistent and correct.

File: tools/bringup/phase2_generate_tests.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    # Confine the (possibly user-supplied) manifest path to the repo root before
+    # reading it, to prevent path traversal outside the intended scope.
+    resolved = os.path.abspath(os.path.join(_REPO_ROOT, os.fspath(manifest_path)))
+    if not resolved.startswith(_REPO_ROOT):
+        raise ValueError(f"manifest path escapes repo root {_REPO_ROOT!r}: {manifest_path!r} (resolved {resolved})")
+    data = json.loads(Path(resolved).read_text(encoding="utf-8"))
+    return list(data.get("records", []))
+
+
+def _select_record_ids(records: Sequence[Dict[str, Any]], only: Optional[List[str]]) -> List[int]:
```
Comment (blozano-tt) [inline tools/bringup/phase2_generate_tests.py:44]: When `--only` matches no records, `_select_record_ids` returns `[]` and the generated test has zero parametrized items. A typo like `--only Conv2D` produces a green pytest run that tested nothing. Consider raising `SystemExit` here if `only` is set but `record_ids` is empty, listing the available kinds.

File: tools/bringup/tracer_test_harness.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    score = pcc(y_nchw, y_ref2)
+    if score < float(pcc_threshold):
+        raise AssertionError(
+            f"PCC={score:.6f} < {float(pcc_threshold):.6f} "
+            f"(rec_id={record.idx}, kind={kind}, in_shape={record.in_shape}, out_shape={record.out_shape})"
+        )
+    return score
+
+
+def run_record_id(
```
Comment (blozano-tt) [inline tools/bringup/tracer_test_harness.py:322]: `run_record_id` calls `load_manifest(manifest_path)` for every parametrized test case. For N records, the manifest is parsed and rebuilt N times. Consider caching the parsed manifest (e.g. `functools.lru_cache` on `load_manifest` or a fixture) to avoid O(N²) overhead.

File: tools/bringup/tracer_op_specs.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        required_params: ``params`` keys the harness dereferences to replay the
+            op; the validator checks these are present. These are exactly the
+            extra ``params`` a given op (like ``Conv2d``) reads at runtime.
+        uses_weight: Whether the op carries a weight artifact (``w_path``).
+        uses_bias: Whether the op carries a bias artifact (``b_path``).
+        runnable: Whether ``tracer_test_harness`` can replay this kind on device.
+    """
+
+    kind: str
+    required_params: Tuple[str, ...] = ()
```
Comment (blozano-tt) [inline tools/bringup/tracer_op_specs.py:65]: `uses_weight` and `uses_bias` are declared on every `OpSpec` but appear to be never read anywhere. This means the validator does not enforce that weight/bias artifacts exist for ops that require them, so a bad manifest can still reach the harness. Either consume these fields in `shared_validate_record` or remove them to avoid implying a contract that is not enforced.

File: tools/bringup/trace_manifest_validation.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+
+
+def _default_shape_loader(path: Path) -> Tuple[int, ...]:
+    """Load a ``.pt`` artifact and return its tensor shape (mirrors the harness)."""
+    import torch
+
+    obj = torch.load(path, map_location="cpu")
+    if isinstance(obj, torch.Tensor):
+        return tuple(obj.shape)
+    if isinstance(obj, dict):
```
Comment (blozano-tt) [inline tools/bringup/trace_manifest_validation.py:84]: When a `.pt` file is a dict, this loader returns the shape of the first tensor via dict iteration. If the dict contains multiple tensors, this is arbitrary and may validate the wrong shape while reporting success. Consider requiring a bare tensor or a well-known key, and erroring on ambiguous multi-tensor dicts.

File: tools/bringup/tracer_op_specs.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+    dilation = _as_pair(params.get("dilation"))
+    if None in (kernel, stride, padding, dilation):
+        errors.append("Conv2d 'params' kernel_size/stride/padding/dilation must be ints or length-2 int pairs")
+        return errors
+
+    for dim, (in_sz, out_sz, k, s, p, d) in (
+        ("height", (h_in, h_out, kernel[0], stride[0], padding[0], dilation[0])),
+        ("width", (w_in, w_out, kernel[1], stride[1], padding[1], dilation[1])),
+    ):
+        if s <= 0:
```
Comment (blozano-tt) [inline tools/bringup/tracer_op_specs.py:363]: The conv2d consistency check validates `stride > 0` but does not guard `padding >= 0` or `dilation >= 1`. Negative padding or zero/negative dilation could pass the output-size formula and reach the device, failing late. Add the same cheap guard here.

File: tools/bringup/phase1_record_ops.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    if any(k.startswith("module.") for k in sd.keys()):
+        return {k[len("module.") :]: v for k, v in sd.items()}
+    return sd
+
+
+def build_model_from_ckpt(ckpt_path: Path | None) -> nn.Module:
+    # Matches your train.py model construction: norm="group", gn_groups=16, bridge_kernel_size=1, bilinear upsample by default.
+    model = UNetVGG19(
+        num_classes=1,
+        pretrained=True,  # you train with pretrained_encoder=True in config :contentReference[oaicite:2]{index=2}
```
Comment (blozano-tt) [inline tools/bringup/phase1_record_ops.py]: These comments contain `:contentReference[oaicite:2]{index=2}` markers and second-person LLM prose. Please clean them up before merging.

**Comment by blozano-tt** [comment]
## Summary Review

Overall this is a solid, well-documented feature. The shared-schema design in  (consumed by both the preflight validator and the runtime harness) is the right architectural choice, and the test coverage in  is strong and well-organized.

**Blockers / significant issues to address:**
1. **Bypassable path guard** (, , ) — the  prefix check can be escaped by a sibling directory. Use the existing  helper based on  everywhere.
2. **O(N²) manifest re-parsing** () —  reloads the manifest for every parametrized test case. Cache it with  or load once in a fixture.
3. **Silent zero-test generation** () — when  matches no records, pytest reports success with nothing tested. Fail loudly if the selection is empty.
4. **Dead weight/bias contract** () — / are declared but never read, so the validator does not enforce that required weight/bias artifacts exist. Consume them in  or remove them.

**Moderate / polish items noted inline:**
- Ambiguous multi-tensor  dict shape loading.
- Missing / guards in conv2d consistency check.
-  coerced to the string  in .
- Dead  variable in generated test files.
- LLM citation artifacts in  comments.
- Bare  swallowing in  and override parsing.

Once the path guard and the empty-selection issue are fixed, this should be good to land.

**Comment by blozano-tt** [comment]
## Summary Review

Overall this is a solid, well-documented feature. The shared-schema design in `tools/bringup/tracer_op_specs.py` (consumed by both the preflight validator and the runtime harness) is the right architectural choice, and the test coverage in `tools/bringup/test_validate_trace_manifest.py` is strong and well-organized.

**Blockers / significant issues to address:**
1. **Bypassable path guard** (`tools/bringup/tracer_op_specs.py:215`, `tools/bringup/phase2_generate_tests.py:33`, `tools/bringup/phase1_record_ops.py:129`) — the `startswith` prefix check can be escaped by a sibling directory. Use the existing `resolve_within_repo` helper based on `os.path.commonpath` everywhere.
2. **O(N²) manifest re-parsing** (`tools/bringup/tracer_test_harness.py:310`) — `run_record_id` reloads the manifest for every parametrized test case. Cache it with `functools.lru_cache` or load once in a fixture.
3. **Silent zero-test generation** (`tools/bringup/phase2_generate_tests.py:55`) — when `--only` matches no records, pytest reports success with nothing tested. Fail loudly if the selection is empty.
4. **Dead weight/bias contract** (`tools/bringup/tracer_op_specs.py:70`) — `uses_weight`/`uses_bias` are declared but never read, so the validator does not enforce that required weight/bias artifacts exist. Consume them in `shared_validate_record` or remove them.

**Moderate / polish items noted inline.**

Once the path guard and the empty-selection issue are fixed, this should be good to land.

File: tools/bringup/tracer_op_specs.py (hunk 1 above)
Comment (ctr-martemovTT) [reply tools/bringup/tracer_op_specs.py]: **resolved**: updated the code to to fix the prefix bug. The helper was not used as it was previously rejected by the the security linter, so I update the individual checks in place instead.

File: tools/bringup/phase2_generate_tests.py (hunk 2 above)
Comment (ctr-martemovTT) [reply tools/bringup/phase2_generate_tests.py]: **resolved**: updated to use the separator

File: tools/bringup/phase2_generate_tests.py (hunk 3 above)
Comment (ctr-martemovTT) [reply tools/bringup/phase2_generate_tests.py:44]: **resolved**: updated the check to raise on empty record ids list

File: tools/bringup/trace_manifest_validation.py (hunk 6 above)
Comment (ctr-martemovTT) [reply tools/bringup/trace_manifest_validation.py:84]: **resolved**: added multi-tensor extract and raise to the ambiguous tensor, and the

File: tools/bringup/tracer_test_harness.py (hunk 4 above)
Comment (ctr-martemovTT) [reply tools/bringup/tracer_test_harness.py:322]: **resolved**: added lru cache

File: tools/bringup/tracer_op_specs.py (hunk 5 above)
Comment (ctr-martemovTT) [reply tools/bringup/tracer_op_specs.py:65]: **resolved**: removed use bias annotation from the conv2d ops, added logic to the shared validate record. The set of ops for tracing is going to be expanded in the phase 3, so in the future other operations will rely on these fields.
